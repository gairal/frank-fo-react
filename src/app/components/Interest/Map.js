import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GoogleMap from 'react-google-maps/lib/components/GoogleMap';
import Marker from 'react-google-maps/lib/components/Marker';
import { withScriptjs, withGoogleMap } from 'react-google-maps/lib';

const GMap = withScriptjs(
  withGoogleMap(({ children, center, zoom }) => (
    <GoogleMap zoom={zoom} center={center}>
      {children}
    </GoogleMap>
  )),
);

export default class Map extends Component {
  static propTypes = {
    travels: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      markers: [],
      center: { lat: 48.856149, lng: 2.3386383 },
      zoom: 8,
    };

    this.isAlive = true;
  }

  componentDidUpdate() {
    const { travels } = this.props;
    const { markers } = this.state;
    if (travels.length && !markers.length) {
      this.travel(0);
    }
  }

  componentWillUnmount() {
    this.isAlive = false;
  }

  travel(index) {
    const { travels } = this.props;
    if (this.isAlive) {
      const currTravel = travels[index];
      if (!currTravel) {
        this.setState(prevState => ({
          ...prevState,
          zoom: 2,
        }));

        return;
      }

      const travel = {
        lat: currTravel.coordinates.Lat,
        lng: currTravel.coordinates.Lng,
      };

      this.setState(prevState => ({
        ...prevState,
        markers: prevState.markers.concat(travel),
        center: travel,
      }));

      setTimeout(() => this.travel(index + 1), 2000);
    }
  }

  render() {
    const { markers, center, zoom } = this.state;
    return (
      <GMap
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyALPF75dwSUZIWeAM-eBvMNA4BN3p-cCj4"
        loadingElement={<div style={{ height: '100%' }} />}
        containerElement={<div style={{ height: '400px' }} />}
        mapElement={<div style={{ height: '100%' }} />}
        center={center}
        zoom={zoom}
      >
        {markers.map(e => (
          <Marker key={e.lat} position={e} />
        ))}
      </GMap>
    );
  }
}
