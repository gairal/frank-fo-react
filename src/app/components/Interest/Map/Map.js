import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GoogleMap from 'react-google-maps/lib/components/GoogleMap';
import Marker from 'react-google-maps/lib/components/Marker';
import { withScriptjs, withGoogleMap } from 'react-google-maps/lib';

const GMap = withScriptjs(withGoogleMap(({ children, center, zoom }) => (
  <GoogleMap
    zoom={zoom}
    center={center}
  >
    { children }
  </GoogleMap>
)));

export default class Map extends Component {
  static propTypes = {
    travels: PropTypes.arrayOf(PropTypes.object).isRequired,
  }

  constructor(props) {
    super(props);

    this.state = {
      markers: [],
      center: { lat: 48.856149, lng: 2.3386383 },
      zoom: 8,
    };

    this.isAlive = true;
  }

  componentDidUpdate() {
    if (this.props.travels.length && !this.state.markers.length) {
      this.travel(0);
    }
  }

  componentWillUnmount() {
    this.isAlive = false;
  }

  travel(index) {
    if (this.isAlive) {
      const currTravel = this.props.travels[index];
      if (!currTravel) {
        this.setState(prevState => ({
          ...prevState,
          zoom: 2,
        }));

        return;
      }

      const travel = {
        lat: currTravel.coordinates.latitude,
        lng: currTravel.coordinates.longitude,
      };

      this.setState(prevState => ({
        ...prevState,
        markers: prevState.markers.concat(travel),
        center: travel,
      }));

      setTimeout(() => this.travel(index + 1), 2000);
    }
  }

  render() {
    return (
      <GMap
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyALPF75dwSUZIWeAM-eBvMNA4BN3p-cCj4"
        loadingElement={<div style={{ height: '100%' }} />}
        containerElement={<div style={{ height: '400px' }} />}
        mapElement={<div style={{ height: '100%' }} />}
        travels={this.props.travels}
        center={this.state.center}
        zoom={this.state.zoom}
      >
        {this.state.markers.map(e => (
          <Marker
            key={e.lat}
            position={e}
          />
        ))}
      </GMap>
    );
  }
}
