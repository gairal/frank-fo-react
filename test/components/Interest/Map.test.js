import React from 'react';
import renderer from 'react-test-renderer';
import Map from '@/components/Interest/Map';

describe('components > Interest > Map', () => {
  jest.useFakeTimers();
  const component = renderer.create(<Map
    travels={[
      {
        place: 'test',
        coordinates: {
          Lat: 0.12,
          Lng: 1.3,
        },
      },
    ]}
  />);

  it('Renders correctly default', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Will update its markers', () => {
    component.root.instance.travel(0);
    expect(component.root.instance.state.zoom).toEqual(8);
    expect(component.root.instance.state.markers).toEqual([{
      lat: 0.12,
      lng: 1.3,
    }]);

    component.root.instance.travel(1);
    expect(component.root.instance.state.zoom).toEqual(2);

    component.unmount();
  });
});
