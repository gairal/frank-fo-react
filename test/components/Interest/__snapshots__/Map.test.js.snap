// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Interest > Map Renders correctly default 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;
